cli/src/commands/git/push.rs

```diff
@@ -275,9 +276,16 @@ pub fn cmd_git_push(
     if old_heads.is_empty() {
         old_heads.push(repo.store().root_commit_id().clone());
     }
-    for commit in revset::walk_revs(repo.as_ref(), &new_heads, &old_heads)?
-        .iter()
-        .commits(repo.store())
+    // (old_heads | immutable())..new_heads
+    let commits_to_push = RevsetExpression::commits(old_heads)
+        .union(&revset_util::parse_immutable_expression(
+            &tx.base_workspace_helper().revset_parse_context(),
+        )?)
```

Comment:
nit: might be better to extract `parse_immutable_heads_expression()` that doesn't call `.ancestors()`.

Reply:
Done.